src/server.rs

```diff
@@ -0,0 +1,450 @@
+    // handler.
+
+    if token.is_err() {
+        return Err(token.err().unwrap());
+    }
+    let max_index = DatabaseManager::max_index(&conn, &user_id, &device_id, &service);
+    let index = options.index.unwrap_or(0 as i64);
+    let limit = options.limit.unwrap_or(0 as i64);
+    let messages =
+        DatabaseManager::read_records(&conn, &user_id, &device_id, &service, index, limit).unwrap();
+    // returns json {"status":200, "index": max_index, "messages":[{"index": #, "data": String}, ...]}
+    let mut is_last = true;
+    if limit > 0 && messages.len() == limit as usize {
+        let last = messages.last().unwrap();
+        is_last = last.idx == max_index;
+    }
+    Ok(Json(json!({
+        "last": is_last,
+        "index": max_index.clone(),
+        "status": 200,
+        "messages": messages
+    })))
+}
+
+#[get("/<service>/<user_id>/<device_id>")]
+fn read(
+    conn: Conn,
+    token: HandlerResult<FxAAuthenticator>,
+    service: String,
+    user_id: String,
+    device_id: String,
+) -> HandlerResult<Json> {
+    // 👩🏫 note that the "token" var is a HandlerResult wrapped Validate struct.
+    // Validate::from_request extracts the token from the Authorization header, validates it
+    // against FxA and the method, and either returns OK or an error. We need to reraise it to the
+    // handler.
+
+    if token.is_err() {
+        return Err(token.err().unwrap());
+    }
+    let max_id = DatabaseManager::max_index(&conn, &user_id, &device_id, &service);
+    // returns json {"status":200, "index": max_index, "messages":[{"index": #, "data": String}, ...]}
+    let messages = match DatabaseManager::read_records(&conn, &user_id, &device_id, &service, 0, 0)
+    {
+        Ok(val) => val,
+        Err(e) => return Err(e),
+    };
+    Ok(Json(json!({
+        "last":true,
+        "index": max_id.clone(),
+        "status": 200,
+        "messages": messages
+    })))
+}
+
+/// Write the user data to the database.
+#[post("/<service>/<user_id>/<device_id>", data = "<data>")]
+fn write(
+    conn: Conn,
+    config: RustboxConfig,
+    token: HandlerResult<FxAAuthenticator>,
+    service: String,
+    user_id: String,
+    device_id: String,
+    data: Json<DataRecord>,
+) -> HandlerResult<Json> {
+    if token.is_err() {
+        return Err(token.err().unwrap());
+    }
+    if config
+        .test_data
+        .get("auth_only")
+        .unwrap_or(&config::Value::from(false))
+        .as_bool()
+        .unwrap_or(false)
+    {
+        // Auth testing, do not write to db.
+        println!("INFO: Auth Skipping database check.");
+        return Ok(Json(json!({
+            "status": 200,
+            "index": -1,
+        })));
+    }
+
+    let response = DatabaseManager::new_record(
+        &conn,
+        &user_id,
+        &device_id,
+        &service,
+        &data.data,
+        calc_ttl(data.ttl),
+    );
+    if response.is_err() {
```

Comment:
same here and in delete, you can replace these 3 lines by using the ? operator on the previous statement